src/providers/hana/qgshanaconnection.cpp

```diff
@@ -0,0 +1,621 @@
+      layerProperty.tableComment = QgsHanaUtils::toQString( rsViews->getNString( 6 ) );
+      layerProperty.isView = true;
+
+      addNewLayer( list, layerProperty, layerProperty.geometryColName.isEmpty() );
+    }
+    rsViews->close();
+
+    int size = list.size();
+    for ( int i = 0; i < size; ++i )
+    {
+      QgsHanaLayerProperty &lp1 = list[i];
+      bool found = false;
+      for ( int j = 0; j < size; ++j )
+      {
+        if ( i != j )
+        {
+          const QgsHanaLayerProperty &lp2 = list.at( j );
+          if ( lp1.schemaName == lp2.schemaName && lp1.tableName == lp2.tableName )
+          {
+            found = true;
+            break;
+          }
+        }
+      }
+
+      if ( !found )
+        lp1.isUnique = true;
+    }
+  }
+  catch ( const Exception &ex )
+  {
+    throw QgsHanaException( ex.what() );
+  }
+  return list;
+}
+
+void QgsHanaConnection::readLayerInfo( QgsHanaLayerProperty &layerProperty )
+{
+  layerProperty.srid = getLayerSRID( layerProperty );
+  layerProperty.type = getLayerGeometryType( layerProperty );
+  layerProperty.pkCols = getLayerPrimaryeKeys( layerProperty );
+}
+
+QVector<QgsHanaSchemaProperty> QgsHanaConnection::getSchemas( const QString &ownerName )
+{
+  QVector<QgsHanaSchemaProperty> list;
+  QString sql = QStringLiteral( "SELECT SCHEMA_NAME, SCHEMA_OWNER FROM SYS.SCHEMAS WHERE "
+                                "HAS_PRIVILEGES = 'TRUE' AND SCHEMA_OWNER LIKE '%1' AND "
+                                "SCHEMA_NAME NOT LIKE_REGEXPR 'SYS|_SYS.*|UIS|SAP_XS|SAP_REST|HANA_XS|XSSQLCC_' AND "
+                                "SCHEMA_NAME LIKE '%2'" )
+                .arg( ownerName.isEmpty() ? QStringLiteral( "%" ) : ownerName, mUri.schema().isEmpty() ? QStringLiteral( "%" ) : mUri.schema() );
+  try
+  {
+    StatementRef stmt = mConnection->createStatement();
+    ResultSetRef rsSchemas = stmt->executeQuery( reinterpret_cast<const char16_t *>( sql.unicode() ) );
+    while ( rsSchemas->next() )
+    {
+      QgsHanaSchemaProperty schema;
+      schema.name = QgsHanaUtils::toQString( rsSchemas->getNString( 1 ) );
+      schema.owner = QgsHanaUtils::toQString( rsSchemas->getNString( 2 ) );
+      list << schema;
+    }
+    rsSchemas->close();
+  }
+  catch ( const Exception &ex )
+  {
+    throw QgsHanaException( ex.what() );
+  }
+  return list;
+}
+
+int QgsHanaConnection::getLayerSRID( const QgsHanaLayerProperty &layerProperty )
+{
+  if ( layerProperty.geometryColName.isEmpty() )
+    return -1;
+
+  int ret = -1;
+
+  try
+  {
+    StatementRef stmt = mConnection->createStatement();
+
+    if ( !layerProperty.isView )
+    {
+      QString sql = QStringLiteral( "SELECT SRS_ID FROM SYS.ST_GEOMETRY_COLUMNS "
+                                    "WHERE SCHEMA_NAME = '%1' AND TABLE_NAME = '%2' AND COLUMN_NAME = '%3'" )
+                    .arg( layerProperty.schemaName, layerProperty.tableName, layerProperty.geometryColName );
```

Comment:
Same as above: non-prepared statements are bad for the database's statement cache and require manual escaping of string literals, which is missing here.

Reply:
Done.